### PR TITLE
multisig:sign command

### PR DESCRIPTION
**Pull request by patnir**
## Summary

This command performs the commitment signature, commitment aggregate, signature create and signature aggregate.

It also checks if the user wants to broadcast the transaction, watch the transaction, etc.

## Documentation

Does this change require any updates to the Iron Fish Docs (ex. [the RPC API
Reference](https://ironfish.network/docs/onboarding/rpc/chain))? If yes, link a
related documentation pull request for the website.

```
[ ] Yes
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and label it with `breaking-change-rpc` or `breaking-change-sdk`.

```
[ ] Yes
```

### CODE REVIEW

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+import { IronfishCommand } from '../../../command'
+import { RemoteFlags } from '../../../flags'
+import * as ui from '../../../ui'
+import { Ledger } from '../../../utils/ledger'
+import { renderUnsignedTransactionDetails, watchTransaction } from '../../../utils/transaction'
+
+// todo(patnir): this command does not differentiate between a participant and an account. 
+// there is a possibility that the account and participant have different names. 
+export class SignMultisigTransactionCommand extends IronfishCommand {
+  static description = 'Interactive command to create a  a multisig transaction'
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:17]: nit - `sign a transaction with a multisig account`?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+      char: 'a',
+      description: 'Name of the account to use for signing the transaction',
+    }),
+    ledger: Flags.boolean({
+      default: false,
+      description: 'Perform operation with a ledger device',
+    }),
+  }
+
+  async collectStrings(
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:35]: should this be a `ui` util?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    const withoutDuplicates = [...new Set(result)]
+
+    if (withoutDuplicates.length !== result.length) {
+      throw new Error(`Duplicate ${item} found in the list`)
+    }
+
+    return result
+  }
+
+  private async retryStep<T>(stepFunction: () => Promise<T>): Promise<T> {
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:58]: should this be a util?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:60]: do you think it's worthwhile to have a maximum number of retries?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    this.log(commitment)
+    this.log('\n============================================')
+
+    const willAggregateCommitments = await ui.confirmPrompt(
+      'Do you want to aggregate the commitments? (Needs to be performed by 1 participant)',
+    )
+
+    let signingPackage: string
+
+    if (willAggregateCommitments) {
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:137]: I don't think there's any randomization in generating signing packages from commitments, so it should be fine if every participant aggregates commitments

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+        unsignedTransaction,
+        unsignedTransactionInput,
+        ledger,
+      )
+    })
+
+    this.log('\n============================================')
+    this.log('\nCommitment:')
+    this.log(commitment)
+    this.log('\n============================================')
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:129]: should we also include an instruction to share the commitment?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        signingPackage,
+        unsignedTransaction,
+        ledger,
+      ),
+    )
+
+    this.log('\n============================================')
+    this.log('\nSignature Share:')
+    this.log(signatureShare)
+    this.log('\n============================================')
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:170]: same comment re: message about sharing the signature share

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+    const willAggregateSignatures = await ui.confirmPrompt(
+      'Do you want to aggregate the signatures? (Needs to be performed by 1 participant)',
+    )
+
+    if (willAggregateSignatures) {
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:176]: It's also ok if everyone aggregates signature shares

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+
+  private async performAggregateSignatures(
+    client: RpcClient,
+    participantName: string,
+    signingPackage: string,
+    signatureShare: string,
+    totalParticipants: number,
+  ): Promise<void> {
+    const broadcast = await ui.confirmPrompt('Do you want to broadcast the transaction?')
+    const watch = await ui.confirmPrompt('Do you want to watch the transaction?')
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:203]: I think these prompts should happen after the prompts to enter signature shares

File: ironfish-cli/src/commands/wallet/multisig/sign.ts [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+    const signingPackageResponse = await client.wallet.multisig.createSigningPackage({
+      account: participantName,
+      unsignedTransaction: unsignedTransaction.serialize().toString('hex'),
+      commitments,
+    })
+
+    return signingPackageResponse.content.signingPackage
+  }
+
+  private async performRound1(
```
Comment (hughy) [inline ironfish-cli/src/commands/wallet/multisig/sign.ts:313]: `performCreateSigningCommitment`?

File: ironfish-cli/src/commands/wallet/multisig/sign.ts (hunk 6 above)
Comment (patnir) [reply ironfish-cli/src/commands/wallet/multisig/sign.ts:98]: done

File: ironfish-cli/src/commands/wallet/multisig/sign.ts (hunk 8 above)
Comment (patnir) [reply ironfish-cli/src/commands/wallet/multisig/sign.ts]: sounds good, removed

File: ironfish-cli/src/commands/wallet/multisig/sign.ts (hunk 5 above)
Comment (patnir) [reply ironfish-cli/src/commands/wallet/multisig/sign.ts]: removed

**Review by hughy — APPROVED** [review]

**Pull request merged**